File: changes/windows-webgpu-service/harness/tools/browser-probe-server.py
#!/usr/bin/env python3
"""Tiny harness server: bind 127.0.0.1:<port>, accept one POST to /, write the
body to <out>, then exit. Used by the webgpu-browser-probe CI workflow to
capture JSON from validate-probe.html without scraping the live DOM.

Exits 0 on report received, 2 on timeout, 3 on misuse.
"""
from __future__ import annotations

import argparse
import http.server
import pathlib
import sys
import threading

class Handler(http.server.BaseHTTPRequestHandler):
    server_version = "webgpu-probe-harness/1.0"

    def log_message(self, fmt, *args):
        # Route access log to stderr with a consistent prefix.
        sys.stderr.write("[harness] " + (fmt % args) + "\n")

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_POST(self):
        length = int(self.headers.get("Content-Length", "0") or "0")
        body = self.rfile.read(length) if length > 0 else b""
        out = self.server.out_path  # type: ignore[attr-defined]
        try:
            out.write_bytes(body)
        except OSError as e:
            self.send_response(500)
            self._cors()
            self.end_headers()
            self.wfile.write(str(e).encode("utf-8", "replace"))
            return
        self.send_response(200)
        self._cors()
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(b'{"ok":true,"bytes":%d}' % len(body))
        print(f"[harness] received {len(body)} bytes -> {out}", flush=True)
        # Tell the server thread to stop, then the main thread to exit.
        self.server.received = True  # type: ignore[attr-defined]
        threading.Thread(target=self.server.shutdown, daemon=True).start()

def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--port", type=int, default=8787)
    ap.add_argument("--out", default="report.json")
    ap.add_argument("--timeout", type=int, default=60)
    args = ap.parse_args()

    out_path = pathlib.Path(args.out).resolve()
    if not (0 < args.port < 65536):
        print("invalid --port", file=sys.stderr)
        return 3

    server = http.server.HTTPServer(("127.0.0.1", args.port), Handler)
    server.out_path = out_path  # type: ignore[attr-defined]
    server.received = False  # type: ignore[attr-defined]
    print(f"[harness] listening on http://127.0.0.1:{args.port}/  "
          f"(timeout {args.timeout}s)", flush=True)

    t = threading.Thread(target=server.serve_forever, daemon=True)
    t.start()
    t.join(timeout=args.timeout)

    if server.received:  # type: ignore[attr-defined]
        return 0
    server.shutdown()
    print(f"[harness] timed out after {args.timeout}s waiting for report",
          file=sys.stderr)
    return 2

if __name__ == "__main__":
    sys.exit(main())

File: changes/windows-webgpu-service/harness/tools/browser-probe-server.js
#!/usr/bin/env node
//
// Tiny harness server: binds to 127.0.0.1:<port>, accepts one POST to /,
// writes the request body to <out>, then exits. CI uses this as a
// side-channel to capture JSON from validate-probe.html without scraping
// the live DOM.
//
// Usage:
//   node browser-probe-server.js --port 8787 --out report.json [--timeout 60]
//
// Exits 0 when a report is received, 2 on timeout, 3 on misuse.

'use strict';
const http = require('http');
const fs = require('fs');
const path = require('path');

function parseArgs(argv) {
    const out = { port: 8787, out: 'report.json', timeout: 60 };
    for (let i = 2; i < argv.length; i++) {
        const a = argv[i];
        if (a === '--port')         out.port    = parseInt(argv[++i], 10);
        else if (a === '--out')     out.out     = argv[++i];
        else if (a === '--timeout') out.timeout = parseInt(argv[++i], 10);
        else if (a === '-h' || a === '--help') { console.log(__filename); process.exit(0); }
        else { console.error(`unknown arg: ${a}`); process.exit(3); }
    }
    if (!Number.isFinite(out.port) || out.port <= 0 || out.port > 65535) {
        console.error(`invalid --port`); process.exit(3);
    }
    return out;
}

const args = parseArgs(process.argv);
const outPath = path.resolve(args.out);
let gotReport = false;

const server = http.createServer((req, res) => {
    // CORS preflight from fetch().
    if (req.method === 'OPTIONS') {
        res.writeHead(204, {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Max-Age': '86400',
        });
        res.end();
        return;
    }
    if (req.method !== 'POST') {
        res.writeHead(405, { 'Allow': 'POST' });
        res.end();
        return;
    }

    let chunks = [];
    req.on('data', c => chunks.push(c));
    req.on('end', () => {
        const buf = Buffer.concat(chunks);
        try {
            fs.writeFileSync(outPath, buf);
        } catch (e) {
            console.error(`write failed: ${e.message}`);
            res.writeHead(500);
            res.end();
            return;
        }
        // Reply with CORS headers so fetch() completes cleanly.
        res.writeHead(200, {
            'Access-Control-Allow-Origin': '*',
            'Content-Type': 'application/json',
        });
        res.end(JSON.stringify({ ok: true, bytes: buf.length }));
        console.log(`[harness] received ${buf.length} bytes → ${outPath}`);
        gotReport = true;
        // Close gracefully after the response flushes.
        server.close(() => process.exit(0));
    });
});

server.on('error', (e) => {
    console.error(`[harness] server error: ${e.message}`);
    process.exit(3);
});

server.listen(args.port, '127.0.0.1', () => {
    console.log(`[harness] listening on http://127.0.0.1:${args.port}/  (timeout ${args.timeout}s)`);
});

setTimeout(() => {
    if (!gotReport) {
        console.error(`[harness] timed out after ${args.timeout}s waiting for report`);
        process.exit(2);
    }
}, args.timeout * 1000).unref();
